common/src/pack.rs

```diff
@@ -0,0 +1,176 @@
+///
+/// When using this, one should consider the performance impact of using
+/// Serde and the associated serializer when targeting BPF. The state of this
+/// is not entirely clear as of now.
+#[macro_export]
+macro_rules! packable {
+    ($my_struct:ty) => {
+        use serum_common::pack::Pack;
+        use solana_client_gen::solana_sdk::program_error::ProgramError;
+
+        impl<'a> Pack<'a> for $my_struct {
+            fn pack(src: $my_struct, dst: &mut [u8]) -> Result<(), ProgramError> {
+                if src.size()? != dst.len() as u64 {
+                    return Err(ProgramError::InvalidAccountData);
+                }
+                serum_common::pack::into_bytes(&src, dst)
+            }
+            fn unpack_unchecked(src: &[u8]) -> Result<$my_struct, ProgramError> {
+                serum_common::pack::from_bytes::<$my_struct>(src)
+            }
+            fn size(&self) -> Result<u64, ProgramError> {
+                serum_common::pack::bytes_size(&self)
+            }
+        }
+    };
+}
+
+pub fn to_bytes<T: ?Sized>(i: &T) -> Result<Vec<u8>, ProgramError>
+where
+    T: serde::Serialize,
+{
+    bincode::serialize(i).map_err(|_| ProgramError::InvalidAccountData)
+}
+
+pub fn into_bytes<T: ?Sized>(i: &T, dst: &mut [u8]) -> Result<(), ProgramError>
+where
+    T: serde::Serialize,
+{
+    let cursor = std::io::Cursor::new(dst);
+    bincode::serialize_into(cursor, i).map_err(|_| ProgramError::InvalidAccountData)
+}
+pub fn from_bytes<'a, T>(data: &'a [u8]) -> Result<T, ProgramError>
+where
+    T: serde::de::Deserialize<'a>,
+{
+    bincode::deserialize(data).map_err(|_| ProgramError::InvalidAccountData)
+}
+
+pub fn bytes_size<T: ?Sized>(value: &T) -> Result<u64, ProgramError>
+where
+    T: serde::Serialize,
+{
+    bincode::serialized_size(value).map_err(|_| ProgramError::InvalidAccountData)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate as serum_common;
+
+    #[derive(Clone, Debug, Default, PartialEq, serde::Serialize, serde::Deserialize)]
+    struct TestStruct {
+        a: u64,
+        b: u64,
+    }
+
+    packable!(TestStruct);
+
+    #[test]
+    fn pack_unpack() {
+        let strct = TestStruct::default();
+        let mut dst = vec![0; strct.size().unwrap() as usize];
+        TestStruct::pack(strct.clone(), &mut dst).unwrap();
+        let new_strct = TestStruct::unpack(&dst).unwrap();
+        assert_eq!(strct, new_strct)
+    }
+
+    #[test]
+    fn unpack_too_small() {
+        let data = vec![0; 8];
+        let result = TestStruct::unpack(&data);
+        match result {
+            Ok(_) => panic!("expect error"),
+            Err(e) => assert_eq!(e, ProgramError::InvalidAccountData),
+        }
+    }
+
+    #[test]
+    fn unpack_too_large() {
+        let data = vec![0; 100];
+        let result = TestStruct::unpack(&data);
+        match result {
+            Ok(_) => panic!("expect error"),
+            Err(e) => assert_eq!(e, ProgramError::InvalidAccountData),
+        }
```

Comment:
This block can be `assert_eq!(e.unwrap_err(), ProgramError::InvalidAccountData)`, same goes for the next 3 instances.